### ISSUE TITLE
Error handling inconsistent with other Workers

### ISSUE THREAD

**Issue by scottpledger** [w3c/ServiceWorker#1519]
From what I've been able to tell, there's no way to get any kind of information about errors from outside the ServiceWorker's context.  This is made more confusing by the fact that `ServiceWorker` instances extend from `AbstractWorker` and thus have an `onerror` handler which, as far as I can tell, never fires.

While such a feature would not necessarily be helpful during a navigation-triggered update, it would be very useful for initial registration/installation and any updates triggered by calling `update()` on the appropriate `ServiceWorkerRegistration` object.  This could help, for example, when trying to diagnose why a ServiceWorker's state changes in an unexpected way.

Since there are cases where a ServiceWorker can be terminated prior to it being able to send/store errors (for example if the SW OOMs, the browser times out handling of `install` and/or `activate` events, etc.), being able to capture such errors from outside the ServiceWorker's scope is the only way a developer might discover such issues.

**Comment by wanderview** [comment]
@jakearchibald do you recall why we don't bubble errors to the service worker context onerror handler or to the ServiceWorker.onerror handler?  It is a bit confusing that we define ServiceWorker.onerror, but I don't think we ever fire an event there.

**Comment by jakearchibald** [comment]
Yeah, this looks like an oversight. `onerror` was added in the very early days of service worker, but it doesn't look like it was properly designed.

**Comment by jakearchibald** [comment]
Ah wait, I misread this. For a while we had an unused `onerror` endpoint in the spec, but we got rid of it ages ago.

We don't really define `ServiceWorker.onerror`, we just get it from `AbstractWorker`. Our behaviour here is the same as [shared workers](https://html.spec.whatwg.org/multipage/workers.html#runtime-script-errors-2), but that doesn't mean it's great.

The problem is that there's a 1-1 relationship between dedicated workers and `Worker` instances. With shared/service workers, an error event would be more of a broadcast, and it isn't clear if that's a good thing.

**Comment by wanderview** [comment]
What is the concern with broadcast?  Performance?  Ideally errors are not high rate.

Is there an alternative to solve Scott's use case of building a system to monitor service workers for unexpected errors?

**Comment by wanderview** [comment]
@scottpledger regarding the broadcast to all ServiceWorker objects, though... Would it cause you problems if you got the same error to multiple tabs of your site?  Would that cause you to double-report the same error?  Maybe firing on the service worker context onerror would be better and then let your own js logic route to the appropriate place.

**Comment by jakearchibald** [comment]
I guess I'm more thinking "I wonder if there was a good reason this wasn't done for shared worker".

> Is there an alternative to solve Scott's use case of building a system to monitor service workers for unexpected errors?

> Maybe firing on the service worker context onerror would be better and then let your own js logic route to the appropriate place.

Hah, I was just typing something along those lines.

**Comment by scottpledger** [comment]
@wanderview That wouldn't cause problems so long as they didn't bubble up to the main scope's `onerror` handler.  Mainly, I'm just trying to capture errors from an installing/activating SW when the update for it is triggered manually through the `ServiceWorkerRegistration.update()` method.  I honestly think that using the existing `ServiceWorker.onerror` handler is the easiest solution because I don't really care if the error is coming from an already activated ServiceWorker (in the particular case I'm trying to debug), and it's also, to me, the most intuitive place to look.

The problem with firing on the `ServiceWorkerGlobalScope.onerror` handler is in cases where the SW OOMs or has some other fatal error that would kill the context without allowing for the handler to either a) execute or b) complete - there's no way to actually get that data out of the SW scope to any kind of logging mechanism.
